File: dp/isSubsequence.cpp
class Solution {
public:

    bool sol(string s, string t, string ans, int i){
        
        // brute

        if(i>t.size()) return false;

        if(ans==s) return true;


        ans+=t[i];
        if(sol(s, t, ans, i+1) == true) return true;

        ans.pop_back();
        if(sol(s, t, ans, i+1) == true) return true;

        return false;
    }

    bool isSubsequence(string s, string t) {
        string ans;
        // return sol(s,t,ans,0, p);
        // optimal
        int i=0,j=0;

        while(i<t.size() && j<s.size()){
            if(s[j]==t[i]){
                j++;
                i++;
            } else {
                i++;
            }
        }

        if(j==s.size()) return true;
        return false;
    }
};

File: dp/isSubsequence.py
class Solution:

    def sol(self, s: str, t: str, ans: str, i: int)-> bool:
        if i>=len(t):
            return ans==s

        if ans==s:
            return True

        ans+=t[i]

        if self.sol(s,t,ans,i+1) == True:
            return True

        ans=ans[:-1]
        if self.sol(s,t,ans,i+1) == True:
            return True

        return False


    def isSubsequence(self, s: str, t: str) -> bool:
    # brute
        
        return self.sol(s, t, "", 0)

    # optimal
        # i=0
        # j=0

        # while i<len(s) and j<len(t):
        #     if s[i]==t[j]:
        #         i+=1
        #         j+=1
        #     else:
        #         j+=1

        # if i==len(s):
        #     return True
        # else:
        #     return False
    
        